test/services/database/challenge_repository_test.dart

```diff
@@ -0,0 +1,206 @@
+        inChallengeRange,
+        ChallengeRepositoryService.maxActiveChallenges,
+      );
+
+      final challenges = await challengeRepository.getChallenges(uid, userPos);
+
+      expect(challenges.length, ChallengeRepositoryService.maxActiveChallenges);
+      for (final challenge in challenges) {
+        expect(challenge.data.isCompleted, false);
+        expect(challenge.data.expiresOn.compareTo(Timestamp.now()), isPositive);
+
+        final actualPost = await postRepository.getPost(challenge.postId);
+        expect(fakePosts.contains(actualPost.data), true);
+      }
+    });
+
+    test("Complete a challenge", () async {
+      await addPosts(postRepository, inChallengeRange, 1);
+      final challenges = await challengeRepository.getChallenges(uid, userPos);
+
+      final challenge = challenges.first;
+      expect(challenge.data.isCompleted, false);
+      await challengeRepository.completeChallenge(
+        uid,
+        challenge.postId,
+      );
+
+      final updatedChallenges =
+          await challengeRepository.getChallenges(uid, userPos);
+      expect(updatedChallenges.length, 1);
+      expect(updatedChallenges.first.data.isCompleted, true);
+    });
+
+    test("Multiple gets with only one challenge available", () async {
+      await addPosts(postRepository, inChallengeRange, 1);
+
+      for (int i = 0; i < 10; i++) {
+        final challenges =
+            await challengeRepository.getChallenges(uid, userPos);
+
+        expect(challenges.length, 1);
+      }
+    });
+
+    test("Challenge posts are unique", () async {
+      await addPosts(
+        postRepository,
+        inChallengeRange,
+        ChallengeRepositoryService.maxActiveChallenges,
+      );
+
+      final challenges = await challengeRepository.getChallenges(
+        uid,
+        userPos,
+      );
+      final postIds = challenges.map((e) => e.postId).toSet();
+      expect(postIds.length, challenges.length);
+    });
+
+    test("Deleted post disappears from challenges", () async {
+      final fakePosts = PostDataGenerator.generatePostData(1);
+      final postId =
+          await postRepository.addPost(fakePosts.first, inChallengeRange);
+
+      final challenges = await challengeRepository.getChallenges(uid, userPos);
+      expect(challenges.length, 1);
+
+      await postRepository.deletePost(postId);
+      final updatedChallenges =
+          await challengeRepository.getChallenges(uid, userPos);
+
+      expect(updatedChallenges.length, 0);
+    });
+
+    test("Challenges expire", () async {
+      const int totalChallenges =
+          2 * ChallengeRepositoryService.maxActiveChallenges - 1;
+      const int challengesToExpire =
+          ChallengeRepositoryService.maxActiveChallenges;
+
+      await addPosts(
+        postRepository,
+        inChallengeRange,
+        totalChallenges,
+      );
+      final now = DateTime.now();
+      final past =
+          now.subtract(ChallengeRepositoryService.maxChallengeDuration);
```

Comment:
I feel this might be a bit dangerous to be exactly at `now - maxChallengeDuration`. I therefore propose instead:
```suggestion
      final past = now.subtract(
        ChallengeRepositoryService.maxChallengeDuration +
            const Duration(seconds: 1),
      );
```

Reply:
The actual duration subtracted is irrelevant here. What is important is that past < now. We could probably get away with using now since getting the challenges is done a bit later. Do you think I should use another duration like 1 second to prevent confusion ?

Reply:
Oh, yes ok indeed. Yes, I think using one second would make more sense than using the unrelated maxChallengeDuration! :)